src/goto-analyzer/show_callsites.cpp

```diff
@@ -0,0 +1,101 @@
+/*******************************************************************\
+
+Module: Write a function call graph to a CSV file
+
+Author: Norbert Manthey, nmanthey@amazon.de
+
+\*******************************************************************/
+
+#include "show_callsites.h"
+
+#include <goto-programs/compute_called_functions.h>
+#include <goto-programs/goto_model.h>
+#include <util/cprover_prefix.h>
+#include <util/json.h>
+#include <util/prefix.h>
+
+#include <fstream>
+#include <iostream>
+#include <sstream>
+
+/** show_callsites writes a call graph to CSV
+ *
+ * This function creates a map between functions and consumed functions, and
+ * stores this relationship in a CSV file, to be processed outside of
+ * goto-analyzer lateron.
+ *
+ * Functions, whose address is consumed, are listed in a separate section so
+ * that outside reasoning has visibility as well.
+ **/
+
+void show_callsites(
+  const goto_modelt &goto_model,
+  const std::string &output_filename)
+{
+  // generate all functions which we cannot be sure where they are called
+  // because their address is consumed, and stored into function pointers
+  std::unordered_set<irep_idt> excluded_functions;
+  compute_address_taken_functions(
+    goto_model.goto_functions, excluded_functions);
+
+  std::ofstream output_file;
+
+  if(output_filename != "-")
+  {
+    output_file.open(output_filename);
+    if(!output_file.is_open())
+      throw "failed to open callsite output file";
```

Comment:
Please, throw `std::runtime_error` or something similar, rather than a string. We often throw strings to signal an assertion (which is an antipattern, we have `INVARIANT` for that), but in this case, it'd make more sense to display a pretty message to the user, rather than a scary stack trace.